Result/Option layout guarantee clarifications

RalfJung opened the pull request:
- It seems worth spelling out that this guarantee allows particular transmutes.
- After the `Result` section was written, the `Option` section it referenced gained *more* guarantees, saying that `None` is represented as `[0u8; N]`. Those guarantees were not meant to apply to `Result`. Make the `Result` section more self-contained to make this more clear.
- "Type has no fields" is unclear since there is no general definition of what the fields of some arbitrary type are. Replace that by a more accurate description of the actual check implemented [here](https://github.com/rust-lang/rust/blob/e379c7758667f900aaf5551c4553c7d4c121e3e1/compiler/rustc_lint/src/types.rs#L828-L838).

r? @traviscross

RalfJung committed:
Result/Option layout guarantee clarifications

RalfJung committed:
clarify 'no fields'

RalfJung commented on library/core/src/result.rs:
```diff
@@ -230,24 +230,30 @@
+//! all of the following conditions:
 //! * Is a zero-sized type with alignment 1 (a "1-ZST").
-//! * Has no fields.
+//! * Is either a struct/tuple with no fields, or an enum with no variants.
```
@theemathas argues that this is a breaking change -- arguably, the `!` type, or `[T; 0]`, or enums with a single variant and no fields were allowed by the previous wording.

The new wording matches what actually got implemented in https://github.com/rust-lang/rust/pull/122253. That PR also used the "has no fields" wording in a doc comment, which (as it turns out) doesn't actually match what is implemented. @Lokathor then copied it to the docs in https://github.com/rust-lang/rust/pull/124870, and later I copied that wording into the FCP [here](https://github.com/rust-lang/rust/pull/130628#issuecomment-2365492463).

RalfJung replied:
It is also worth mentioning that the [ABI docs](https://doc.rust-lang.org/nightly/std/primitive.fn.html#abi-compatibility) actually document a stronger guarantee: if someone defines their own `enum Result`, they get this guarantee whenever the other type is a 1-ZST, without any further caveats.

This is based on the FCP [here](https://github.com/rust-lang/rust/pull/130628#issuecomment-2402761599). The Result docs never got updated to match that. Should they? And what about the lint? The reason the lint works like this is semver; the ABI docs are deliberately not semver-aware since one may want to use them within one's own crate.

Lokathor replied:
If the compiler supports/checks the more permissive form given in the FCP, then I think the docs should match.

Lokathor replied:
(and if the compiler doesn't support the FCP's wording then the compiler should be updated, followed by a docs update)

RalfJung replied:
> Real Question: is an array a struct/tuple, or is it an enum?

Neither.

@traviscross so the real question here is this one I think:
> The Result docs should either match the FFI-types lint or the ABI guarantees, and currently they do neither, so we definitely want to change something here.

Seems like you're leaning towards having it match the ABI guarantees, so I'll update the PR for that. (I won't be updating the FFI-types lint here, that lint needs a redesign since it's currently not even clear when and how it should take semver into consideration, among other issues.)

Lokathor replied:
If an array is neither then are we deliberately excluding 0-sized arrays from these docs? Or should the wording also permit them?

RalfJung replied:
The FFI lint considers 0-zised 1-aligned arrays to not qualify for the guarantee. I cannot tell whether that is deliberate or not.

Lokathor replied:
Oh, yeah, I was looking in comments for the lint it seems. No problem then, I guess.

RalfJung committed:
have Result docs match ABI docs

RalfJung commented:
@traviscross was it okay to assign you as reviewer here? This isn't meant to say anything new, just to clarify that when we say 
> [`Option<T>`] has the same size, alignment, and [function call ABI] as `T`

then you can transmute between these two types, and that will preserve the values.

theemathas commented:
Does this interact with https://github.com/rust-lang/rust/pull/147185?

RalfJung commented:
I don't see how?

`Result<NonZeroI32, ReprCZst>` is indeed ABI-compatible with `NonZeroI32` if `ReprCZst` truly is a ZST; we can do that since the ABI for this `Result` instance doesn't have to have anything to do with that for `ReprCZst`. Note that we already don't worry about the ZST being non-exhaustive or so: whether this type actually always is a ZST is your problem as unsafe code author relying on this guarantee.

This is quite different from `repr(transparent)` where the compiler checks whether the conditions are met, and where `#[repr(transparent)] struct S((), ReprCZst)` can be interpreted as being compatible with either type if we aren't careful.

traviscross committed:
Revise `Result`/`Option` guarantee docs

The notation used here (e.g. "transmute `t: T` to `Option<T>`") felt
maybe a bit heavy, in the context of the library documentation, so
let's elaborate this a bit.

Also, in the `Option` docs, we talk about this being true for "the
following types `T`", but it felt this caveat might get a bit lost in
the next sentence that talks about the valid transmutations, so let's
reiterate the caveat.

While we're touching the line, we can improve:

> The only difference is the implied semantics:

This sentence was a bit awkward due to the mismatched plurality and
not identifying the difference being spoken to.  We'll reword this to
make it more clear.

We'll wrap to 80, since the existing text and most of the doc comments
in these files are wrapped this way.

bors commented:
:pushpin: Commit 73f5fe78d4b822124ca6082d49592314eb9d68a9 has been approved by `traviscross`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=traviscross 73f5fe78d4b822124ca6082d49592314eb9d68a9 -->
<!-- homu: {"type":"Approved","sha":"73f5fe78d4b822124ca6082d49592314eb9d68a9","approver":"traviscross","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
